HIP/hip-904-frictionless-airdrop.md

```diff
@@ -0,0 +1,711 @@
+
+This can be a significant security concern. A token with custom fees could  be used by a token spammer to airdrop a token to all users with a custom fee that will charge the users some HBAR or other tokens in order to get rid of the token — essentially creating a “ransom” token. To prevent this, we introduce the new `TokenRejectTransaction`.
+
+Any token (fungible or non-fungible) can be rejected, whether it was created through automatic or manual association, and no matter how long the token was held. A rejected token is transferred back to the token treasury, and all custom fees on the token are waived. The payer of the transaction will only pay the appropriate fees for the HAPI transaction type.
+
+> ⚠️ When a token is rejected, we **do not** decrement `used_auto_associations`. That field tracks the total number of times auto-associations have been made, not the current number of token types on the account that were auto-associated.
+
+
+> ⚠️ `receiver_sig_required` is ignored on the treasury account when the `TokenReject` is handled. Reject will always work with no assessed custom fees.
+
+### Example
+
+A company called GameCo wants to airdrop tokens (e.g., unique NFTs to commemorate a specific event) to everyone who attended a game. They can send those NFTs (say, GameNFT) to every attendee without explicit signatures from the thousands of attendees. GameCo will pay all required fees to perform this airdrop. Alice, Bob, Carol, and Dave all attended this game.
+
+Alice has an account with unlimited auto-associations. When GameCo drops the GameNFT into her account, it is immediately associated and transferred. Alice now owns a GameNFT.
+
+Bob has an account with a fixed number of auto-associations, but the number of auto-associations he’s received is less than his fixed max, so he has an open slot. When GameCo drops the GameNFT into his account, it is immediately associated and transferred. Bob now owns a GameNFT.
+
+Carol has an account with auto-associations disabled. She has no open slots. When GameCo drops the GameNFT into her account, she does not immediately receive it. Her wallet alerts her to a new airdrop. She decides she wants this token. For a nominal fee, she signs a transaction to accept this airdrop. Her account is associated with GameNFT and a serial is transferred to her account. Carol now owns a GameNFT.
+
+Dave also has an account with no open slots. When GameCo drops the GameNFT into his account, he does not immediately receive it. His wallet alerts him to a new airdrop, but Dave doesn’t want it. He tells his wallet to ignore this airdrop. A couple days later, Dave regrets his decision. He checks, and the airdrop is still available, so he decides to claim it. If he wanted to, he could have ignored this airdrop until GameCo canceled the airdrop.
```

Comment:
"He tells his wallet to ignore this airdrop" but doesn't trigger a "token airdrop reject" transaction ?
If I have rejected the airdrop via a transaction, presumably it's no longer in my list of pending airdrops and I can't later accept it ?

Reply:
Right. I would like to ignore a solicitation via my wallet without having to spend money to reject it.

Reply:
Note, this is a suggestion of wallet behaviour. The "ignore this airdrop" has no effect on the ledger - no transaction is sent, just logic to ignore a listing.
The only way to remove a pending transfer from state is for the sender to cancel or the receiver to claim.
Neither can impact or dictate the others state change

Reply:
I've added a diagram to show that TokenReject operates only on tokens in an accounts balance. It does not operate on pending token transfers

Reply:
Doesn't it make sense the receivers to be also able to make token rejections even on pending airdrops. The receiver would have to pay for this but it's up to them to make this decision. Having token rejections only on tokens in your balance seems a bit restrictive. A receiver might want to clean their pending state from time to time and dapps or wallets could provide this functionality.

Reply:
The concept is a fair consideration, but if you consider the pending state to be on the sender then it wouldn't be right to allow the receiver to change that when the sender is paying for the state.
If you consider the pending state shared then there's an argument.
The hope is tools will be a bit smart and provide appropriate UX for users to ignore or filter certain pending airdrops from their view.

Also if anything it would be a `TokenCancelAirdrop` for pending transfers.
`TokenReject` only refers to in balance tokens.

Reply:
That's a good argument. It won't be fair a receiver to reject directly an airdrop that has been paid by the sender with all of its fees. Agree on that.